### PR TITLE
4658 add correspondence backend

### PR DESCRIPTION
**Pull request by rachelschneiderman**
Added proxy, lambda, interactor, and Correspondence entity necessary for creating a new correspondence and associating it with a case. Update and delete still WIP.

### CODE REVIEW

File: shared/src/business/entities/cases/Case.js [hunk 1]
```diff
@@ -1751,5 +1767,17 @@ Case.prototype.getCaseConfirmationGeneratedPdfFileName = function () {
   return `case-${this.docketNumber}-confirmation.pdf`;
 };
 
+/**
+ * generates the case confirmation pdf file name
```
Comment (matthopson) [inline shared/src/business/entities/cases/Case.js]: I think this needs to be updated

File: shared/src/business/entities/cases/Case.js [hunk 2]
```diff
@@ -1751,5 +1767,17 @@ Case.prototype.getCaseConfirmationGeneratedPdfFileName = function () {
...
+ *
+ * @param {Correspondence} correspondenceEntity the correspondence document to add to the case
+ * @returns {Case} this case entity
+ */
+Case.prototype.fileCorrespondence = function (correspondenceEntity) {
```
Comment (matthopson) [inline shared/src/business/entities/cases/Case.js:1776]: Is this a real entity?

File: shared/src/business/useCases/correspondence/fileCorrespondenceDocumentInteractor.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+const { Correspondence } = require('../../entities/Correspondence');
+const { UnauthorizedError } = require('../../../errors/errors');
+
+/**
+ *
+ * @param {object} providers the providers object
+ * @param {object} providers.applicationContext the application context
+ * @param {object} providers.documentMetadata the document metadata
+ * @param {string} providers.primaryDocumentFileId the id of the primary document
+ * @returns {Promise<*>} the updated case entity after the document is added
```
Comment (matthopson) [inline shared/src/business/useCases/correspondence/fileCorrespondenceDocumentInteractor.js]: Nit: Probably "raw case object" is more correct

File: shared/src/business/useCases/correspondence/fileCorrespondenceDocumentInteractor.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    {
+      ...documentMetadata,
+      documentId: primaryDocumentFileId,
+      filedBy: user.name,
+      userId: user.userId,
+    },
+    { applicationContext },
+  );
+
+  caseEntity.fileCorrespondence(correspondenceEntity);
```
Comment (matthopson) [inline shared/src/business/useCases/correspondence/fileCorrespondenceDocumentInteractor.js:57]: I see, it is an entity

File: shared/src/proxies/correspondence/updateCorrespondenceDocumentProxy.js [hunk 5]
```diff
@@ -0,0 +1,25 @@
+const { put } = require('../requests');
+
+/**
+ * updateCourtIssuedOrderProxy
```
Comment (matthopson) [inline shared/src/proxies/correspondence/updateCorrespondenceDocumentProxy.js]: Description

File: shared/src/proxies/correspondence/fileCorrespondenceDocumentInteractor.js [hunk 6]
```diff
@@ -0,0 +1,26 @@
+const { post } = require('../requests');
+
+/**
+ * addCorrespondenceDocumentProxy
```
Comment (matthopson) [inline shared/src/proxies/correspondence/fileCorrespondenceDocumentInteractor.js]: `fileCorrespondenceDocumentInteractor`

File: web-client/src/presenter/actions/UploadCorrespondenceDocument/setupUploadMetadataAction.js [hunk 7]
```diff
@@ -0,0 +1,12 @@
+import { state } from 'cerebral';
+
+/**
+ * update the upload form for submission
+ *
+ * @param {object} providers the providers object
+ * @param {object} providers.store the cerebral store
+ * @param {object} providers.get the get function
+ */
+export const setupUploadMetadataAction = ({ get, store }) => {
```
Comment (matthopson) [inline web-client/src/presenter/actions/UploadCorrespondenceDocument/setupUploadMetadataAction.js]: Nit: I think a name like `setFormDocumentTitleFromFreeTextAction` (or maybe something better than that, haha) is a little more truthful here.

File: web-client/src/presenter/actions/UploadCorrespondenceDocument/submitCorrespondenceAction.js [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+import { state } from 'cerebral';
+
+/**
+ * sets the correspondence document onto the case
+ *
+ * @param {object} providers the providers object
+ * @param {object} providers.applicationContext the application context
+ * @param {object} providers.props the cerebral props object
+ * @returns {Promise} async action
+ */
```
Comment (matthopson) [inline web-client/src/presenter/actions/UploadCorrespondenceDocument/submitCorrespondenceAction.js:12]: Missing `get` from jsdocs

File: web-client/src/presenter/actions/UploadCorrespondenceDocument/uploadCorrespondenceFileAction.js [hunk 9]
```diff
@@ -0,0 +1,31 @@
+import { state } from 'cerebral';
+
+/**
+ * upload correspondence document to s3.
+ *
+ * @param {object} providers the providers object
+ * @param {object} providers.applicationContext the application context
+ * @returns {object} the next path based on if validation was successful or error
+ */
```
Comment (matthopson) [inline web-client/src/presenter/actions/UploadCorrespondenceDocument/uploadCorrespondenceFileAction.js:11]: Missing `get`, `path` from jsdocs

File: shared/src/business/entities/Correspondence.js [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+  filingDate: joiStrictTimestamp
+    .max('now')
+    .required()
+    .description('Date that this Document was filed.'),
+  userId: joi.string().required(),
+};
+
+joiValidationDecorator(Correspondence, Correspondence.schema, {});
+
+module.exports = { Correspondence };
```
Comment (matthopson) [inline shared/src/business/entities/Correspondence.js:34]: We usually would include test with the entity that, at the very least, tests validation

File: shared/src/business/useCases/correspondence/fileCorrespondenceDocumentInteractor.js [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+  if (caseEntity.validate()) {
+    await applicationContext.getPersistenceGateway().fileCaseCorrespondence({
+      applicationContext,
+      caseId,
+      correspondence: correspondenceEntity.validate().toRawObject(),
+    });
+  }
+
+  return caseEntity.toRawObject();
+};
```
Comment (matthopson) [inline shared/src/business/useCases/correspondence/fileCorrespondenceDocumentInteractor.js:68]: Missing test

**Review by matthopson — COMMENTED** [review]
We're missing a number of tests. I realize it's a work in progress, but we've struggled in the past with having too much code going in without tests (and the intention to add them after), which brings overall line coverage down. With such a high coverage expectation, it's not hard to upset those margins :P

File: shared/src/business/entities/Correspondence.js (hunk 10 above)
Comment (mrinsin) [reply shared/src/business/entities/Correspondence.js:34]: on it!

File: shared/src/business/useCases/correspondence/fileCorrespondenceDocumentInteractor.test.js [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+  //     applicationContext,
+  //     caseId: 'c54ba5a9-b37b-479d-9201-067ec6e335bb',
+  //     documentId: 'c54ba5a9-b37b-479d-9201-067ec6e335bc',
+  //   });
+
+  //   expect(
+  //     applicationContext.getUseCaseHelpers().sendServedPartiesEmails,
+  //   ).toHaveBeenCalled();
+  //   expect(result).toBeUndefined();
+  // });
```
Comment (matthopson) [inline shared/src/business/useCases/correspondence/fileCorrespondenceDocumentInteractor.test.js]: Is the commented stuff, functionality being worked on?

Comment (matthopson) [reply shared/src/business/useCases/correspondence/fileCorrespondenceDocumentInteractor.test.js]: JK, I see it was removed.. (I'm reviewing by commit, which is cumbersome in that way)

**Review by matthopson — APPROVED** [review]
Good work ya'll!!

**Pull request merged**